Feature/59180 rescheduling on changing week days and holidays

toy opened the pull request:
# Ticket

[OP#63901](https://community.openproject.org/wp/63901) (part of [OP#59180](https://community.openproject.org/wp/59180), follows #18523 and #18824)

# What are you trying to accomplish?
Automatic rescheduling of phases when working days and holidays are change in admin interface

## Screenshots

# What approach did you choose and why?

# Merge checklist

- [ ] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

toy committed:
empty Projects::Phases::ApplyWorkingDaysChangeJob

toy committed:
enque both ApplyWorkingDaysChangeJobs

toy committed:
check for both ApplyWorkingDaysChangeJobs

toy committed:
extract ProjectLifeCycleSteps::RescheduleService

toy committed:
Project::Phase.covering_dates_and_days_of_week adapted from WorkPackage with hackyily adapted test

toy committed:
extract commond code for changed_days and changed_non_working_dates

toy committed:
refactor method for producing a hash with differenc between two lists

Choosing combination of speed and clarity of intent:
```rb
require 'benchmark/ips'

require 'active_support/core_ext/enumerable'

def change_between(list_a, list_b)
  set_a = Set.new(list_a)
  set_b = Set.new(list_b)

  (set_a ^ set_b).index_with { set_b.include?(it) }
end

def change_between1(list_a, list_b)
  was = list_a.to_h { [it, false] }
  now = list_b.to_h { [it, true] }

  now.merge(was){ nil }.compact
end

def change_between2(list_a, list_b)
  deleted = (list_a - list_b).to_h { [it, false] }
  added = (list_b - list_a).to_h { [it, true] }

  deleted.merge(added)
end

def change_between3(list_a, list_b)
  set_a = Set.new(list_a)
  set_b = Set.new(list_b)
  deleted = (set_a - set_b).to_h { [it, false] }
  added = (set_b - set_a).to_h { [it, true] }
  deleted.merge(added)
end

def change_between4(list_a, list_b)
  was = list_a.index_with(false)
  now = list_b.index_with(true)

  now.merge(was){ nil }.compact
end

def change_between5(list_a, list_b)
  deleted = (list_a - list_b).index_with(false)
  added = (list_b - list_a).index_with(true)

  deleted.merge(added)
end

def change_between6(list_a, list_b)
  set_a = Set.new(list_a)
  set_b = Set.new(list_b)
  deleted = (set_a - set_b).index_with(false)
  added = (set_b - set_a).index_with(true)
  deleted.merge(added)
end

def compare(list)
  as_and_bs = 1000.times.map do
    [list.sample(rand(list.length)), list.sample(rand(list.length))]
  end

  change_between_methods = private_methods.grep(/^change_between\d?$/)

  as_and_bs.each do |a, b|
    d = change_between(a, b)
    change_between_methods.each do |change_between_method|
      fail "#{change_between_method} returned different result" unless d
== send(change_between_method, a, b)
    end
  end

  Benchmark.ips do |x|
    change_between_methods.each do |change_between_method|
      x.report(change_between_method) { as_and_bs.map{
send(change_between_method, _1, _2) } }
    end

    x.compare!
  end
end

compare(7.times.to_a)
compare(100.times.to_a)
```

```
ruby 3.4.2 (2025-02-15 revision d2930f8e7a) +PRISM [arm64-darwin23]
Warming up --------------------------------------
      change_between    40.000 i/100ms
     change_between1   158.000 i/100ms
     change_between2   170.000 i/100ms
     change_between3    32.000 i/100ms
     change_between4   152.000 i/100ms
     change_between5   148.000 i/100ms
     change_between6    33.000 i/100ms
Calculating -------------------------------------
      change_between    408.732 (± 2.7%) i/s    (2.45 ms/i) -      2.080k in   5.093365s
     change_between1      1.588k (± 0.7%) i/s  (629.69 μs/i) -      8.058k in   5.074251s
     change_between2      1.645k (±12.2%) i/s  (607.90 μs/i) -      8.160k in   5.048118s
     change_between3    318.922 (± 2.8%) i/s    (3.14 ms/i) -      1.600k in   5.021512s
     change_between4      1.477k (± 1.3%) i/s  (676.94 μs/i) -      7.448k in   5.042753s
     change_between5      1.608k (± 8.5%) i/s  (621.98 μs/i) -      7.992k in   5.012654s
     change_between6    331.757 (± 1.8%) i/s    (3.01 ms/i) -      1.683k in   5.074711s

Comparison:
     change_between2:     1645.0 i/s
     change_between5:     1607.8 i/s - same-ish: difference falls within error
     change_between1:     1588.1 i/s - same-ish: difference falls within error
     change_between4:     1477.2 i/s - same-ish: difference falls within error
      change_between:      408.7 i/s - 4.02x  slower
     change_between6:      331.8 i/s - 4.96x  slower
     change_between3:      318.9 i/s - 5.16x  slower

ruby 3.4.2 (2025-02-15 revision d2930f8e7a) +PRISM [arm64-darwin23]
Warming up --------------------------------------
      change_between     4.000 i/100ms
     change_between1    10.000 i/100ms
     change_between2    10.000 i/100ms
     change_between3     4.000 i/100ms
     change_between4     9.000 i/100ms
     change_between5    11.000 i/100ms
     change_between6     4.000 i/100ms
Calculating -------------------------------------
      change_between     45.431 (± 4.4%) i/s   (22.01 ms/i) -    228.000 in   5.029300s
     change_between1    112.354 (± 7.1%) i/s    (8.90 ms/i) -    560.000 in   5.013557s
     change_between2    133.466 (± 4.5%) i/s    (7.49 ms/i) -    670.000 in   5.031395s
     change_between3     44.247 (± 2.3%) i/s   (22.60 ms/i) -    224.000 in   5.065967s
     change_between4    104.443 (± 4.8%) i/s    (9.57 ms/i) -    522.000 in   5.011400s
     change_between5    114.941 (± 6.1%) i/s    (8.70 ms/i) -    583.000 in   5.094324s
     change_between6     46.083 (± 4.3%) i/s   (21.70 ms/i) -    232.000 in   5.047312s

Comparison:
     change_between2:      133.5 i/s
     change_between5:      114.9 i/s - 1.16x  slower
     change_between1:      112.4 i/s - 1.19x  slower
     change_between4:      104.4 i/s - 1.28x  slower
     change_between6:       46.1 i/s - 2.90x  slower
      change_between:       45.4 i/s - 2.94x  slower
     change_between3:       44.2 i/s - 3.02x  slower
```

toy committed:
cache changed_days and changed_non_working_dates

toy committed:
extract common parts of ApplyWorkingDaysChangeJobs

toy committed:
implement Projects::Phases::ApplyWorkingDaysChangeJob

toy committed:
journal change

toy committed:
all active phases should be rescheduled when working days are added

ulferts commented on app/models/project/phases/scopes/covering_dates_and_days_of_week.rb:
```diff
@@ -0,0 +1,88 @@
+    # @param dates Date[] An array of the Date objects.
+    # @param days_of_week number[] An array of the ISO days of the week to
+    #   consider. 1 is Monday, 7 is Sunday.
+    def covering_dates_and_days_of_week(days_of_week: [], dates: [])
```
This method seems to return all the phases which:
* span a certain weekday OR
* span a certain date

So I think the method should rather be named `covering_dates_or_days_of_week`

ulferts committed:
avoid generating a series of days unnecessarily

ulferts committed:
feature spec for project phase rescheduling on configuration change

toy committed:
duration should be nil if date range is not set

toy committed:
fix and cleanup optimized sql for covering_dates_and_days_of_week

toy committed:
don't generate list of dates in covering_dates_and_days_of_week of work packages too

toy committed:
mock view_project_attributes check to allow all projects

Not clear why wasn't it needed before, but doesn't look important to
investigate

toy committed:
test ProjectLifeCycleSteps::RescheduleContract

toy committed:
add test for Projects::Phases::ApplyWorkingDaysChangeJob not running in Settings::WorkingDaysAndHoursParamsContract

toy committed:
add test for scheduling Projects::Phases::ApplyWorkingDaysChangeJob in Settings::WorkingDaysAndHoursUpdateService

toy committed:
use calculate_duration trait in feature spec instead of defaulting to calculating duration

ulferts commented on app/models/work_packages/scopes/covering_dates_and_days_of_week.rb:
```diff
@@ -59,30 +59,25 @@ def where_covers_periods(work_packages_periods_cte, days_of_week, dates)
+          EXISTS (
+            SELECT 1
+            FROM unnest(Array[:dates]::DATE[]) AS target_date
+            WHERE target_date BETWEEN start_date AND end_date
```
👍 for cleaning this up in lockstep with the sibling scope.

toy committed:
test ProjectLifeCycleSteps::RescheduleService

ulferts commented on app/workers/apply_working_days_change_job_base.rb:
```diff
@@ -0,0 +1,82 @@
+  private
+
+  def apply_working_days_change
+    fail NoMethodError, "Must be overridden in subclass"
```
Not overly important but why is a NoMethodError used here instead of a NotImplementedError?

toy replied:
It is incorrect to use `NotImplementedError` in such case, at it is «Raised when a feature is not implemented on the current platform». Defining method and raising same `NoMethodError` error as when it would not be defined is for documenting and giving more info in the raised exception. So it would be better to use that or introduce an `AbstractMethodError` and use it in all other places, but given that this looks like the only place where `NoMethodError` is used instead of `NotImplementedError`, it can be changed later across the codebase if other developers agree

ulferts replied:
Thanks for the clarification. I wasn't aware of that definition so I misused the `NotImplementedError` in the past it seems.

ulferts commented on app/workers/projects/phases/apply_working_days_change_job.rb:
```diff
@@ -0,0 +1,62 @@
+  private
+
+  def apply_working_days_change
+    user = User.current
```
I think this is here uselessly.

toy replied:
It is needed for `RescheduleService`

ulferts replied:
True, I didn't see the connection. I guess it was moved outside of the loop for performance reasons. However, I think the overhead of calling User.current on a call to 

```
ProjectLifeCycleSteps::RescheduleService.new(user: User.current, project:)
```

would be minimal and would increase readability.

ulferts commented on app/workers/projects/phases/apply_working_days_change_job.rb:
```diff
@@ -0,0 +1,62 @@
+
+    Project::Phase
+      .active # TODO: should visible be used?
+      .covering_dates_and_days_of_week(days_of_week:, dates:)
```
Since a user calling this job is an admin, using the `visible` scope would only change the scope to only include active projects. Archived projects would be included as permissions are not active in there, even for admins. 

That might be a worthwhile approach, though, since it would then not change the scheduling of already completed projects. However, if we follow along this path, we would have to reschedule the phases when reactivating a project. That is a use case, that while being a valid one, I currently judge to be of lesser importance so I would currently not support it. I also wouldn't know where to make the cut. Because one could also argue that phases that are scheduled in the past should not be rescheduled. To me however, that argument only is worth to follow up on when we support validity periods for the non working days settings.

TL;DR: Using the `active` scope seems fine to me.

ulferts commented on app/workers/projects/phases/apply_working_days_change_job.rb:
```diff
@@ -0,0 +1,62 @@
+  def apply_working_days_change
+    user = User.current
+
+    applicable_phases.distinct.pluck(:project_id).each do |project_id|
```
The `distinct` IMO is unnecessary since the scope should not return duplicates. On the other hand, adding a `.uniq` after the `pluck(:project_id)` would save unnecessary iterations. 

Here, however, it might be better to use something like 

```
Project.where(id: applicable_phases.select(:project_id)).each do |project|
```

That way, the check for the existence of the project can be be omitted as well.

toy replied:
`applicable_phases` returns a scope for phases, so can return multiple phases for same project and `distinct` applies to the select from `pluck`. Compare SQLs for simpler queries without and with distinct:
```ruby
Project::Phase.pluck(:project_id)
# SELECT "project_phases"."project_id" FROM "project_phases"
```
```ruby
Project::Phase.distinct.pluck(:project_id)
# SELECT DISTINCT "project_phases"."project_id" FROM "project_phases"
```

I went with iterating ids of projects and loading projects later to not load all projects in memory and because slight decrease of performance loading projects one by one should not be very important here. Alternative is probably to use `Project.where(id: applicable_phases.select(:project_id)).find_each`

ulferts replied:
```
Project.where(id: applicable_phases.select(:project_id)).find_each
```

Is what I would work with, then. To me, the extra loading and checking of the project produces noise.

ulferts commented on app/workers/projects/phases/apply_working_days_change_job.rb:
```diff
@@ -0,0 +1,62 @@
+      phases = project.available_phases
+      next if phases.empty?
+
+      from = phases.first.start_date
```
The first phase might have its start date not set but other phases in the same project can still have dates. Because of that, I think it safer to use 

```
from = phases.pluck(:start_date, :finish_date).flatten.compact.min
```

toy replied:
Good point. First I though that `covering_dates_and_days_of_week` should not allow that, but it returns also phases with only one date set if it is on the affected day of week or date. Still it is better to force querying `phases`, as they are needed afterwards anyway

toy replied:
I'm also not sure if it makes sense to consider `finish_date` if `start_date` is not set

ulferts replied:
When not considering the finish_date, something like this can happen (edge case admittedly):

Have a set of phases where one only has its finish date set. 

<img width="906" alt="image" src="https://github.com/user-attachments/assets/74e4e16b-9292-40cf-b7c6-7ed51a90fdbc" />

Add a non working days to the finish date set for the phase. 

Now add a start date to the phase which only had a finish date:

<img width="909" alt="image" src="https://github.com/user-attachments/assets/f3018059-a8c0-4727-860b-2431540ae6e2" />

I am not sure, if this should be prevented by the automatic rescheduling, though. It could set the date to be on the next non working day. But since it requires the user to interact with the phase again to raise the error it might be acceptable to have the invalid date kept. What do you think, @toy ?

toy replied:
I would keep it as is for now, I think updating finish date when phase has no start date may be more unexpected than allowing it to become invalid

ulferts commented on app/services/project_life_cycle_steps/reschedule_service.rb:
```diff
@@ -0,0 +1,81 @@
+    def reschedule_phases(phases:, from:)
+      phases.each do |phase|
+        next unless phase.date_range_set?
+        next unless phase.duration&.positive? # not updated using service that sets duration
```
Currently the caller is responsible to filter out phases that should not be rescheduled. On the other hand, some selection is taking place in here. Why is `active` not filtered for here?

toy replied:
Good point, while now I rely on input to provide only active phases, it is better to add the check to be less confusing in the future. Also it is a reminder that phases are also not checked for order, so if a scope is used that would return them not in definition order, they will be rescheduled in incorrect order, but it felt odd to load definitions here only to ensure order.

ulferts reviewed (changes requested):
The implementation looks solid and well structured. 

I did not manage to run through a complete set of test cases yet but found one edge case that is worth to take a look at.

### Having project phases with a phase in the middle not having a date set

<img width="913" alt="image" src="https://github.com/user-attachments/assets/47f68fc9-63bd-40c7-b107-38d9834760fc" />

When adding a non working day within the time frame spanned by the other phases, the rescheduling will break.

toy committed:
agreed to use simpler active and not visible scope in the job

toy committed:
force querying phases and use first defined start_date to reschedule from

toy committed:
make reschedule service skip inactive phases

Currently input already contains only active phases, but better to check
it explicitly to reduce confusion possibility

toy committed:
rename covering_dates_and_days_of_week to covering_dates_or_days_of_week

toy committed:
simplify for rubocop

toy committed:
test Projects::Phases::ApplyWorkingDaysChangeJob

toy commented:
@ulferts Thank you for thorough review, I either applied changes or replied, so please have a look

ulferts reviewed (approved):
Good work @toy. From my point of view, this is good to merge. 

There is this one edge case still open which I am undecided about but this can either decided to be ignored or be fixed later on. 

So once the CI is green, go ahead and merge 🎉

toy committed:
pass user exclipictly in ApplyWorkingDaysChangeJob

toy committed:
use batch processing insted of iterating project ids

toy committed:
allow rubocop not to fail
